### ISSUE TITLE
Unusable state.

### ISSUE THREAD

**Issue by CyrusTheVirusG** [lxc/lxd#4989]
# Required information

 * Distribution: Ubuntu 
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails: Hangs
   * Kernel version: 4.15.0-33-generic #36-Ubuntu SMP Wed Aug 15 16:00:05 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
   * LXC version: 3.0.1-0ubuntu1~18.04.2 all
   * LXD version: 3.0.1-0ubuntu1~18.04.1
   * Storage backend in use: btrfs

# Issue description

lxc completely locks up, had setup cluster on one node and was going to add others to it.

Now that I read about the config having to be the same, it makes sense how it got that way.

However, if I purge the lxd of a node because I can't get it to join the cluster it should not break the remaining node, that has containers running on it. Luckily, nothing important.

Raft attempts an election which fails, resulting in a denial of service using the lxc commands. Completely hangs. While I realize the implications of making a change when the other nodes can't sync could be catastrophic, potentially, I don't like not being able to execute any commands.

Perhaps, only allow limited commands while the cluster is 'synchronizing'  like lxc exec, stop, start, retstart, launch. Currently a terrible user experience. 

Additionally, instead of completing the addition of the node to the cluster on the first node, wait until the config is verified good after lxd init. We have some out of order operation happening here whereas the node will get added even though you gave it bad values and it fails to init properly.

Uggh to make matters worse you can't even re-init the failed node after rebooting the first node or maybe it was issuing a lxc command to hang either way it's bad.

# Steps to reproduce

 1. Step one
Setup two machines in a cluster, but create an extra storage pool on the first and watch it fail on the second when you don't specify a different source for the new storage pool.
 2. Step two
apt-get purge lxd, on the second node with an incomplete configuration that somehow still got added to the cluster with a bad config.
 3. Step three
issue any lxc command, hangs indefinitely thousands of connections established to lxd ~2,683 on my system.

# Information to attach
Aug 31 15:04:41 -- lxd[1610]: lvl=warn msg="Raft: Election timeout reached, restarting election" t=2018-08-31T15:04:41+0000
Aug 31 15:04:42 -- lxd[1610]: lvl=warn msg="Failed connecting to global database (attempt 6): gRPC grpcConnection failed: context deadline exceeded" t=2018-08-31T15:04:42+0000

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

**Comment by stgraber** [comment]
Ok, so current LXD does not hang indefinitely anymore, instead it'd wait 10s and tell you that your database is unreachable. `lxc cluster` commands do keep working in that state, so you can see the state of your cluster and get it back online, all other commands cannot work since they all rely on the database and the database isn't reachable at that point in time.

Removing a node, should always be done through `lxc cluster remove` prior to wiping/reinstalling a node. If that's not an option, running `lxc cluster remove --force` will let you remove a node without notifying it.


So with those bits out of the way, let's focus on Step one as that's where there may be an issue we're not aware of or haven't fixed yet. Can you explain exactly what you did?
It sounds like you did:
 - `lxd init` on both nodes to create the cluster, possibly configuring an initial local storage pool
 - Ran `lxc storage create some-name some-type --node node1 source=something`
 - Ran `lxc storage create some-name some-type --node node2`
 - Ran `lxc storage create some-name some-type`
 - At which point the storage pool failed to create on node2?

**Comment by freeekanayaka** [comment]
@stgraber small rectification: if the majority of you database nodes is down, not even ```lxc cluster list``` (or other ```lxc cluster``` commands that require the db) will work. The case you probably were thinking of is when some nodes were updated while others were not: the updated nodes will be blocked waiting for the others to be updated, but if you run ```lxc cluster list``` against one of the nodes that are still behind that will work and indicate which nodes are blocked waiting for the other to update.

**Comment by freeekanayaka** [comment]
I think we've now seen enough folks removing nodes a bit lightly that we'll want to update the docs making the difference between ```lxc cluster remove``` and ```lxc cluster remove --force``` even more clear, and warning about the implications of an ```apt-get remove``` or ```snap remove```.

**Comment by stgraber** [comment]
Yeah, that sounds good, want to update doc/clustering.md @freeekanayaka

**Comment by CyrusTheVirusG** [comment]
This was the second node in the cluster and it didn't init properly which is why I was surprised that the original node was affected.

On the first node I had a btrfs storage pool named 'fast' which was located on /dev/sdb1.

On the second node it kept giving me an error about size when trying to configure it and the init failed.

I now realize that I was trying to give it the entire disk instead of the partition which caused it to fail. Works great when you do it correctly but that first time you mess up it's a huge pain to troubleshoot when all the commands lag massively.

 Really a 1-5 second timeout should be more than sufficient, maybe even flag it as down for 15-30 seconds with an exponential backoff timer. 

So that you can issue commands and get an error back immediately which maybe says something to the effect of "Check cluster status using 'lxc cluster list', command processing disabled until quorum re-established..." or something like that...

Otherwise you really have to purge lxd to get the system back to a usable state. Especially if you didn't know the cluster commands would still work. 

I got everything working on my four nodes, now I have an issue with the images server misbehaving... probably a dnsmasq issue or something. Works fine on my other 16.04 boxes.

**Comment by CyrusTheVirusG** [comment]
I should also note that an existing btrfs partition /dev/sdb1 also causes init to fail, instead of overwriting or using the existing partition.

**Comment by stgraber** [comment]
We chose the 10s value on newer LXD as an amount of time a user would likely wait before canceling whatever they're doing while still allowing enough time for temporary database unavailability as this happens when one node gets an upgrade (as auto-upgrades happen for snap users) or when a leader election happens due to network changes.

That way you only really get an error if something is actually wrong and needs attention rather than if the database is just temporarily unavailable because of ongoing upgrade or topology changes.

**Comment by stgraber** [comment]
We'll need to do some tests on 3.0.1 and 3.0.2 to see if the fixes we did on init fix this issue.
I'm reasonably sure that the new init API we have in 3.3 fixes this, but as it wasn't backported to 3.0.2, we need to check and see how 3.0.2 behaves.

**Comment by stgraber** [comment]
@CyrusTheVirusG I'm unable to reproduce your issue. I did:
```
root@edfu:~# lxc storage create blah btrfs --target nuturo source=/dev/sdd
Storage pool blah pending on member nuturo
root@edfu:~# lxc storage create blah btrfs --target edfu source=/dev/sdd
Storage pool blah pending on member edfu
root@edfu:~# lxc storage create blah btrfs
Error: Failed to create the BTRFS pool: /dev/sdd appears to contain an existing filesystem (btrfs).
ERROR: use the -f option to force overwrite of /dev/sdd
btrfs-progs v4.15.1
See http://btrfs.wiki.kernel.org for more information.
```

Which does show the error you'd expect on a busy block device, but the resulting status remains correct:
```
root@edfu:~# lxc storage list
+-------+-------------+--------+---------+---------+
| NAME  | DESCRIPTION | DRIVER |  STATE  | USED BY |
+-------+-------------+--------+---------+---------+
| blah  |             | btrfs  | PENDING | 0       |
+-------+-------------+--------+---------+---------+
| local |             | btrfs  | CREATED | 1       |
+-------+-------------+--------+---------+---------+
```

The pool remains in state "PENDING"  and so the create can be re-attempted once the block device is cleaned up:
```
root@edfu:~# dd if=/dev/zero of=/dev/sdd bs=4M count=10
10+0 records in
10+0 records out
41943040 bytes (42 MB, 40 MiB) copied, 0.579297 s, 72.4 MB/s
root@edfu:~# lxc storage create blah btrfs
Storage pool blah created
```

**Comment by stgraber** [comment]
And to better match your example above:
```
root@edfu:~# lxc storage create blah btrfs --target nuturo source=/dev/sdd
Storage pool blah pending on member nuturo
root@edfu:~# lxc storage create blah btrfs --target edfu size=10GB
Storage pool blah pending on member edfu
root@edfu:~# lxc storage create blah btrfs
Storage pool blah created
```

Note that when skipping the `source` property, it's recommended that you tell LXD what size of loop device you want, which is done with size=10GB above.

**Comment by CyrusTheVirusG** [comment]
The error occurs when using sudo lxd init on the second node. I assume, perhaps incorrectly that creating a storage pool before lxd is init would result in an lxd init failure due to there being an existing config.

**Comment by stgraber** [comment]
```
root@nuturo:~# lxc storage create blah btrfs
Storage pool blah created
root@nuturo:~# lxc storage list
+------+-------------+--------+-----------------------------------------+---------+
| NAME | DESCRIPTION | DRIVER |                 SOURCE                  | USED BY |
+------+-------------+--------+-----------------------------------------+---------+
| blah |             | btrfs  | /var/snap/lxd/common/lxd/disks/blah.img | 0       |
+------+-------------+--------+-----------------------------------------+---------+
root@nuturo:~# lxd init
Would you like to use LXD clustering? (yes/no) [default=no]: yes
What name should be used to identify this node in the cluster? [default=nuturo]: 
What IP address or DNS name should be used to reach this node? [default=172.17.16.50]: 
Are you joining an existing cluster? (yes/no) [default=no]: yes
IP address or FQDN of an existing cluster node: 172.17.16.48
Cluster fingerprint: c4c7beda413be65b343f47a72c9cd4c479e256c8ebd78b643c9a45517868e1e6
You can validate this fingerpring by running "lxc info" locally on an existing node.
Is this the correct fingerprint? (yes/no) [default=no]: yes
Cluster trust password: 
All existing data is lost when joining a cluster, continue? (yes/no) [default=no] yes
Choose "size" property for storage pool "local": 
Choose "source" property for storage pool "local": 
Choose "zfs.pool_name" property for storage pool "local": 
Would you like a YAML "lxd init" preseed to be printed? (yes/no) [default=no]: 
root@nuturo:~# lxc storage list
+-------+-------------+--------+---------+---------+
| NAME  | DESCRIPTION | DRIVER |  STATE  | USED BY |
+-------+-------------+--------+---------+---------+
| local |             | zfs    | CREATED | 1       |
+-------+-------------+--------+---------+---------+
root@nuturo:~# 
```

As `lxd init` mentions, any existing configuration is lost when joining the cluster.
The database of the joining node is effectively wiped clean, it's therefore a bad idea to have anything on a node prior to joining, but as can be seen above, if that's the case, it gets lost.

I'm going to close this issue as it went kind of all over the place with so far no definitive steps to reproduce an actual bug in a supported release of LXD (3.0.2 or 3.5).

If you do have a reproducer for a cluster related issue, comment back here or file a separate issue and we'll look into it.
